service/watcher/chartstatus/chartstatus.go

```diff
@@ -0,0 +1,186 @@
+package chartstatus
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/app/v4/pkg/key"
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/google/go-cmp/cmp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/giantswarm/app-operator/v3/pkg/annotation"
+)
+
+type ChartStatusWatcherConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+
+	ChartNamespace string
+	PodNamespace   string
+	UniqueApp      bool
+}
+
+type ChartStatusWatcher struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+
+	appNamespace   string
+	chartNamespace string
+	uniqueApp      bool
+}
+
+func NewChartStatusWatcher(config ChartStatusWatcherConfig) (*ChartStatusWatcher, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.ChartNamespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ChartNamespace must not be empty", config)
+	}
+	if config.PodNamespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.PodNamespace must not be empty", config)
+	}
+
+	c := &ChartStatusWatcher{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		// We get a kubeconfig for the cluster from the chart-operator app CR
+		// which is in the same namespace as this instance of app-operator.
+		appNamespace:   config.PodNamespace,
+		chartNamespace: config.ChartNamespace,
+		uniqueApp:      config.UniqueApp,
+	}
+
+	return c, nil
+}
+
+func (c *ChartStatusWatcher) Boot(ctx context.Context) {
+	go c.watchChartStatus(ctx)
+}
+
+// watchChartStatus watches all chart CRs in the target cluster for status
+// changes. The matching app CR status is updated otherwise there can be a
+// delay of up to 5 minutes until the next resync period.
+func (c *ChartStatusWatcher) watchChartStatus(ctx context.Context) {
+	for {
+		// We need an active kubeconfig to connect to the cluster as it may be
+		// remote. We get this from the chart-operator app CR. The connection
+		// to the cluster may sometimes be down. So we wait with a backoff
+		// until we can connect.
+		g8sClient, err := c.waitForActiveKubeConfig(ctx)
```

Comment:
At times like on cluster creation or upgrades the workload cluster may not be available. The backoff retries until a connection can be made.